File: src/DshLauncher/Models/LauncherSettings.cs
namespace DshLauncher.Models;

public enum DshInstallScope
{
    Global,
    Managed
}

public sealed class DshInstallation
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "DSH";
    public DshInstallScope Scope { get; set; }
    public string InstallRoot { get; set; } = string.Empty;
    public string PackageRoot { get; set; } = string.Empty;
    public string NodeExecutable { get; set; } = string.Empty;
    public string NpmExecutable { get; set; } = string.Empty;
    public string InstalledVersion { get; set; } = string.Empty;
    public DateTimeOffset? LastVerifiedAt { get; set; }

    public override string ToString()
    {
        var scope = Scope == DshInstallScope.Global ? "\u5168\u5C40" : "\u72EC\u7ACB";
        return string.IsNullOrWhiteSpace(InstalledVersion)
            ? Name + "\uFF08" + scope + "\uFF0C\u672A\u9A8C\u8BC1\uFF09"
            : Name + " " + InstalledVersion + "\uFF08" + scope + "\uFF09";
    }
}

public sealed class DshHomeEntry
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = "\u9ED8\u8BA4\u6570\u636E";
    public string Path { get; set; } = string.Empty;
    public string? LastObservedWriterVersion { get; set; }
    public DateTimeOffset? LastObservedWriteAt { get; set; }
    public bool ObservationReliable { get; set; }

    public override string ToString() => $"{Name} \u00B7 {Path}";
}

public sealed class LauncherSettings
{
    public int SchemaVersion { get; set; } = 1;
    public List<DshInstallation> Installations { get; set; } = [];
    public List<DshHomeEntry> Homes { get; set; } = [];
    public Guid? SelectedInstallationId { get; set; }
    public Guid? SelectedHomeId { get; set; }
    public int Port { get; set; } = 3080;
    public string WorkingDirectory { get; set; } = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
    public bool StartDshWithLauncher { get; set; } = true;
    public bool OpenBrowserAfterStart { get; set; } = true;
    public bool StartWithWindows { get; set; }
    public bool NotifyOnCompletion { get; set; } = true;
    public string LauncherUpdateRepository { get; set; } = "shawnyhu/DSH-Launcher";

    public DshInstallation? SelectedInstallation =>
        Installations.FirstOrDefault(x => x.Id == SelectedInstallationId);

    public DshHomeEntry? SelectedHome =>
        Homes.FirstOrDefault(x => x.Id == SelectedHomeId);
}

public enum DshActivityState
{
    Stopped,
    Idle,
    Busy,
    Attention,
    Incompatible
}

public sealed record DshStatusSnapshot(
    DshActivityState State,
    int RunningSessions,
    int PendingQuestions,
    int PendingApprovals,
    string Summary)
{
    public static readonly DshStatusSnapshot Stopped = new(
        DshActivityState.Stopped, 0, 0, 0, "DSH \u672A\u8FD0\u884C");
}

public sealed record DshNotification(
    string Title,
    string Message,
    ToolTipIcon Icon,
    bool IsCompletion = false);

File: macos/DSHLauncher/Models/LauncherSettings.swift
import Foundation

enum DSHInstallScope: String, Codable, CaseIterable, Sendable {
    case global
    case managed
}

struct DSHInstallation: Codable, Identifiable, Equatable, Sendable {
    var id = UUID()
    var name: String
    var scope: DSHInstallScope
    var installRoot: String
    var packageRoot: String
    var nodeExecutable: String
    var npmExecutable: String
    var installedVersion: String
    var lastVerifiedAt: Date?

    var displayName: String {
        let kind = scope == .global ? "全局" : "独立"
        return "\(name) \(installedVersion)（\(kind)）"
    }
}

struct DSHHomeEntry: Codable, Identifiable, Equatable, Sendable {
    var id = UUID()
    var name: String
    var path: String
    var lastObservedWriterVersion: String?
    var lastObservedWriteAt: Date?
    var observationReliable = false
}

struct LauncherSettings: Codable, Equatable, Sendable {
    static let currentSchemaVersion = 1

    var schemaVersion = currentSchemaVersion
    var installations: [DSHInstallation] = []
    var homes: [DSHHomeEntry]
    var selectedInstallationID: UUID?
    var selectedHomeID: UUID?
    var port = 3080
    var workingDirectory: String
    var startDSHWithLauncher = true
    var openBrowserAfterStart = true
    var startAtLogin = false
    var notifyOnCompletion = true
    var launcherUpdateRepository = "shawnyhu/DSH-Launcher"

    static func defaults(homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser) -> Self {
        let home = DSHHomeEntry(name: "默认数据", path: homeDirectory.appendingPathComponent(".dsh").path)
        return LauncherSettings(
            homes: [home],
            selectedHomeID: home.id,
            workingDirectory: homeDirectory.path
        )
    }

    var selectedInstallation: DSHInstallation? {
        installations.first { $0.id == selectedInstallationID }
    }

    var selectedHome: DSHHomeEntry? {
        homes.first { $0.id == selectedHomeID }
    }

    mutating func selectInstallation(_ id: UUID?) {
        guard let id, installations.contains(where: { $0.id == id }) else { return }
        selectedInstallationID = id
    }

    mutating func selectHome(_ id: UUID?) {
        guard let id, homes.contains(where: { $0.id == id }) else { return }
        selectedHomeID = id
    }

    func normalized(homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser) -> Self {
        var value = self
        if value.homes.isEmpty {
            let home = DSHHomeEntry(name: "默认数据", path: homeDirectory.appendingPathComponent(".dsh").path)
            value.homes = [home]
            value.selectedHomeID = home.id
        } else if value.selectedHome == nil {
            value.selectedHomeID = value.homes[0].id
        }
        if value.selectedInstallation == nil {
            value.selectedInstallationID = value.installations.first?.id
        }
        value.port = min(max(value.port, 1), 65_535)
        if value.workingDirectory.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            value.workingDirectory = homeDirectory.path
        }
        return value
    }
}

enum DSHActivityState: String, Sendable {
    case stopped
    case idle
    case busy
    case attention
    case incompatible
}

struct DSHStatusSnapshot: Equatable, Sendable {
    var state: DSHActivityState
    var runningSessions: Int
    var runningJobs: Int
    var pendingQuestions: Int
    var pendingApprovals: Int
    var summary: String

    static let stopped = DSHStatusSnapshot(
        state: .stopped,
        runningSessions: 0,
        runningJobs: 0,
        pendingQuestions: 0,
        pendingApprovals: 0,
        summary: "DSH 未运行"
    )
}

struct OperationProgress: Equatable, Sendable {
    var stage: String
    var percentage: Int?
    var detail: String?
}
